arlo_server/ballots.py

```diff
@@ -0,0 +1,118 @@
+import io, csv
+from sqlalchemy import func, literal_column
+from sqlalchemy.dialects.postgresql import aggregate_order_by
+
+from arlo_server import app
+from arlo_server.auth import with_jurisdiction_access
+from arlo_server.models import (
+    Election,
+    SampledBallotDraw,
+    Round,
+    Batch,
+    AuditBoard,
+    SampledBallot,
+    Jurisdiction,
+)
+from util.csv_download import csv_response, election_timestamp_name
+
+
+def ballot_retrieval_list(jurisdiction: Jurisdiction, round: Round) -> str:
+    csv_io = io.StringIO()
+    retrieval_list_writer = csv.writer(csv_io)
+    retrieval_list_writer.writerow(
+        [
+            "Batch Name",
+            "Ballot Number",
+            "Storage Location",
+            "Tabulator",
+            "Ticket Numbers",
+            "Already Audited",
+            "Audit Board",
+        ]
+    )
+
+    # Get previously sampled ballots as a separate query for clarity
+    # (self joins are cool but they're not super clear)
+    previous_ballots_query = (
+        SampledBallotDraw.query.join(SampledBallotDraw.round)
+        .filter(Round.round_num < round.round_num)
+        .join(SampledBallotDraw.batch)
+        .filter_by(jurisdiction_id=jurisdiction.id)
+        .values(Batch.name, SampledBallotDraw.ballot_position)
+    )
+    previous_ballots = {
+        (batch_name, ballot_position)
+        for batch_name, ballot_position in previous_ballots_query
+    }
+
+    # Get deduped sampled ballots
+    ballots = (
+        SampledBallotDraw.query.filter_by(round_id=round.id)
+        .join(SampledBallotDraw.batch)
+        .filter_by(jurisdiction_id=jurisdiction.id)
+        .join(SampledBallotDraw.sampled_ballot)
+        .join(SampledBallot.audit_board)
+        .add_entity(Batch)
+        .add_entity(AuditBoard)
+        .group_by(
+            Batch.name,
+            Batch.id,
+            Batch.storage_location,
+            Batch.tabulator,
+            AuditBoard.name,
+        )
+        .group_by(SampledBallotDraw.ballot_position)
```

Comment:
Any reason not to combine the group_by's?

Reply:
I just copy-pasted your code without making any changes. So maybe I'll ask you: any reason not to combine the group_by's? 😝

Reply:
went ahead and did some cleanup: a5bbe58